timApp/idesupport/utils.py

```diff
@@ -228,6 +232,85 @@ class TIDECourse:
     """
 
 
+def combine_code_blocks(program: str | None, by: str | None,
+                        by_code: str | None, filename: str | None) -> str:
+    """
+    Combine the code blocks.
+
+    :param program: Code block
+    :param by: Code block
+    :param by_code: Code block
+    :return: Combined code blocks
+    """
+    if program is None and by is None and by_code is None:
+        raise RouteException("No code found in the plugin")
+
+    if by is None and by_code is None:
+        return program
+
+    # Find the comment line syntax based on the file extension and create messages.
+    comment_line_syntax = find_comment_line_syntax(filename.split(".")[-1])
+    user_code_begins_message = comment_line_syntax + \
+        " --- Write your code below this line. ---"
+    user_code_ends_message = comment_line_syntax + \
+        " --- Write your code above this line. ---"
+
+    # If the program is not meant to be fully editable, add the messages to the code.
+    if program != "REPLACEBYCODE":
+        annotated_block = user_code_begins_message + "\n" + \
+            "REPLACEBYCODE" + "\n" + user_code_ends_message
+        program = program.replace("REPLACEBYCODE", annotated_block)
+
+    combined_code = ""
+
+    if by is not None:
+        combined_code = program.replace("REPLACEBYCODE", by)
+
+    if by_code is not None:
+        combined_code = program.replace("REPLACEBYCODE", by_code)
+
+    return combined_code
+
+
+def find_comment_line_syntax(file_extension: str) -> str:
+    """
+    Find the comment line syntax based on the type.
+
+    :param type: Type of the file
+    :return: Comment line syntax
+    """
+    comment_syntax_lookup = {
+        "cpp": "//",
+        "c": "//",
+        "cs": "//",
+        "java": "//",
+        "py": "#",
+        "js": "//",
+        "html": "<!--",
+    }
+
+    return comment_syntax_lookup.get(file_extension, "//")
+
+
+def ensure_filename(ide_file: IdeFile, task_info: TIDETaskInfo) -> IdeFile:
+    """
+    Ensure that the filename is set for the ide_file.
+
+    :param ide_file: IdeFile
+    :param task_info: TIDETaskInfo
+    :return: IdeFile
+    """
+
+    filename = "main"
+    programming_languages = ["c", "cpp", "cs", "java", "py", "js", "html"]
+    for lang in programming_languages:
+        if lang in task_info.type:
+            filename = "main." + lang
+            break
```

Comment:
Vois siirtää tonne luokan metodiksi.

Bugi pitäisi korjata logiikasta, merkkijonosta ei haluta tässä tarkistaa matchia "in" operaatiolla

Reply:
Muutettu